### ISSUE TITLE
Vertex data and attribute declaration dependency

### ISSUE THREAD

**Issue by vorg** [regl-project/regl#328]
It seems that in regl the vertex attribute has to match the input data dimensions e.g.

```glsl
attribute vec2 position;
```
```javascript
regl({
  attributes: {
    position: [[0, 0], [1, 0], [1, 0]]
  }
})
```

So now when i declare position as having 4 components nothing renders on the screen.

```glsl
attribute vec4 position;
```

According to this [SO Answer](http://stackoverflow.com/questions/8551935/opengl-es-2-0-specifiying-position-attribute-vec3-or-vec4) and OpenGL ES 2.0 Spec the attrbute would be expanded to 4 components with missing data filled as follows `[0, 0, 0, 1]` and this is how i've written my shaders so far with no problems.

How is the attribute size guessed in regl?

**Comment by mikolalysenko** [comment]
It guesses the attribute size using the shader.  You can overload this by setting the `size` property in the attribute description.  For example:

```javascript
attributes: {
  position: {
     size: 4,
     data: [[0, 0], [1, 0], [1, 0]]
  }
}
```

**Comment by vorg** [comment]
As or regl 1.2.1 and vertex shader with the following attribute declaration

```glsl
attribute vec4 position;
```

The example from above

```javascript
attributes: {
  position: regl.buffer({
     size: 4,
     data: [[0, 0], [1, 0], [1, 0]]
  })
}
```
Will throw `Error: (regl) missing buffer for attribute "position" in command unknown`


Switching to 
```javascript
attributes: {
  position: {
     size: 4,
     data: [[0, 0], [1, 0], [1, 0]]
  }
}
```
will throw `WebGL: INVALID_OPERATION: drawArrays: attempt to access out of bounds arrays`

Changing   `position: { size: 4 ...` to `position: { size: 2...` will still throw `WebGL: INVALID_OPERATION: drawArrays: attempt to access out of bounds arrays`

I believe this is invalid behavior as now my geometry data is dependent on shader code.

To sum up:
shader attr size == command attr size -> OK
shader attr size > command attr size -> access out of bounds arrays
shader attr size < command attr size -> black screen

**Comment by mikolalysenko** [comment]
Sorry about that it should be:

```
attributes: {
  position: {
     size: 2,
     data: [[0, 0], [1, 0], [1, 0]]
  }
}
```

**Comment by vorg** [comment]
@mikolalysenko no this doesn't work, i need to create regl.buffer() and match the attribute size in JS and vertex shader

**Comment by vorg** [comment]
Working code: https://gist.github.com/vorg/8ef68bc8fa769c21d3d97ea753147029

**Comment by monfera** [comment]
FWIW I've run into it also, assuming that the size is figured out from the shader reflection as the doc suggests (results of the `gl.getActiveAttrib` call), but indeed, an explicit size specification was needed (2 for `mat2`, 3 for `mat3`, 4 for `mat4`). I was going to make a small PR updating the doc but maybe the spirit of `regl` is to figure out the size and then found this issue.

`attribute mat2 position;`
...
```
      attributes: {

        position: {
          size: 2, // needed
          buffer: regl.buffer([
            0, 0,
            1, 1
          ])
        },
/* // this didn't work:
        position: [
            0, 0,
            1, 1
          ]
*/
      }
```
